src/contracts/core/Slasher.sol

```diff
@@ -1,102 +1,368 @@
+        uint256 slashedEpochHistoryLength = slashedEpochHistory[operator][strategy].length;
+        if (slashedEpochHistoryLength == 0) {
+            // TODO: consider if a different return value is more appropriate for this special case
+            return 0;
+        } else {
+            return slashedEpochHistory[operator][strategy][slashedEpochHistoryLength - 1];
+        }
+    }
 
-    function operatorToMiddlewareTimes(
-        address,
-        uint256
-    ) external view returns (MiddlewareTimes memory) {}
+    function _modifyRequestedBipsToSlash(
+        address operator,
+        bytes4 operatorSetID,
+        IStrategy[] memory strategies,
+        uint32 epoch,
+        int32 bipsToModify
+    ) internal {
+        require(bipsToModify != 0, "Slasher._modifyRequestedBipsToSlash: cannot modify slashing by 0");
 
-    function middlewareTimesLength(address) external view returns (uint256) {}
+        IOperatorSetManager.OperatorSet memory operatorSet =
+            IOperatorSetManager.OperatorSet({avs: msg.sender, id: operatorSetID});
+        bytes32 operatorSetHash = _hashOperatorSet(operatorSet);
+        for (uint256 i = 0; i < strategies.length; ++i) {
+            IStrategy strategy = strategies[i];
+            uint32 requestedSlashedBipsBefore = requestedSlashedBips[operator][strategy][epoch][operatorSetHash];
+            int32 requestedSlashedBipsAfter = int32(requestedSlashedBipsBefore) + bipsToModify;
+            // don't allow underflow
+            if (requestedSlashedBipsAfter < 0) {
+                bipsToModify = -int32(requestedSlashedBipsBefore);
+                requestedSlashedBipsAfter = 0;
+            }
+            requestedSlashedBips[operator][strategy][epoch][operatorSetHash] = uint32(requestedSlashedBipsAfter);
 
-    function getMiddlewareTimesIndexStalestUpdateBlock(address, uint32) external view returns (uint32) {}
+            SlashingRequest memory slashingRequest = slashingRequests[operator][strategy][epoch];
+            // if this is the first request, increment the lastCreatedSlashingRequestId and set the id
+            if (slashingRequest.id == 0) {
+                uint32 requestId = slashingRequestIds[operator][strategy].lastCreatedSlashingRequestId + 1;
+                slashingRequestIds[operator][strategy].lastCreatedSlashingRequestId = requestId;
+                slashingRequest.id = requestId;
+            }
 
-    function getMiddlewareTimesIndexServeUntilBlock(address, uint32) external view returns (uint32) {}
+            // pending slashing rate is modified by the bips to modify * the slashable bips
+            // when divided by BIPS_FACTOR**2, this will give the actual proportion of the
+            // stake is being modified. this is done for accuracy when slashable bips are small
+            slashingRequest.slashingRate = uint64(
+                int64(slashingRequest.slashingRate)
```

Comment:
order of operations looks incorrect here. Need to add bipsToModify to slashingRate first(add parenthesis) before multiplying with the operator's slashable bips/magnitude

Reply:
Nvm the slashingRate is parts per hundred million (BIPS_FACTOR_SQUARED)
The slashingRate is adding the product of the operator's configured magnitude and the requested bips to slash by the AVS and adding this to the existing slashingRate. Makes sense now